File: src/code-intel-interfaces/code_intel_interfaces/analyzer.py
from __future__ import annotations
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any


class AnalyzerType(str, Enum):
    """Known analyzer types. Extensible via string values."""
    GIT_COUPLING = "git"
    DEPENDENCY = "deps"
    SEMANTIC = "semantic"
    INTELLIGENCE = "intelligence"


class TaskStatus(str, Enum):
    NOT_RUN = "not_run"
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class AnalysisTask:
    """A unit of work dispatched to an analyzer."""
    task_id: str
    analyzer_type: str
    repo_id: str
    repo_path: Path          # path to mirror.git or source repo
    db_path: Path            # path to the shared SQLite database
    parquet_dir: Path        # path to parquet data directory
    config: dict[str, Any]   # analyzer-specific configuration


@dataclass
class TaskResult:
    """Standard result from any analyzer task."""
    task_id: str
    status: TaskStatus
    entity_count: int = 0
    relationship_count: int = 0
    metrics: dict[str, Any] = field(default_factory=dict)
    error: str | None = None


class BaseAnalyzer(ABC):
    """
    Contract that every analyzer must implement.

    The orchestrator:
    1. Calls list_capabilities() to discover what this analyzer offers
    2. Calls analyze() to run analysis (async-friendly, status tracked in DB)
    3. Calls the analyzer-specific query interface for data retrieval
    """

    @property
    @abstractmethod
    def analyzer_type(self) -> str:
        """Unique identifier: 'git', 'deps', 'semantic', etc."""
        ...

    @property
    @abstractmethod
    def display_name(self) -> str:
        """Human-readable name for UI."""
        ...

    @abstractmethod
    def get_config_schema(self) -> dict:
        """JSON Schema for this analyzer's configuration."""
        ...

    @abstractmethod
    def analyze(self, task: AnalysisTask) -> TaskResult:
        """
        Run analysis. This is the main entry point.

        SYNC tasks: Return TaskResult directly with data.
        ASYNC tasks: Update status in DB, return TaskResult with status=RUNNING,
                     complete work in background.

        The analyzer writes entities and relationships to the shared DB
        using the standard schema.
        """
        ...

    def validate_config(self, config: dict) -> list[str]:
        """Validate configuration. Return list of error messages (empty = valid)."""
        return []

File: src/frontend/src/types/analyzer.ts
// Analyzer task and status types

export interface AnalyzerInfo {
    type: AnalyzerType;
    name: string;
    description: string;
    available: boolean;
    last_run?: string;
    status?: TaskStatus;
}

export type AnalyzerType = 'git' | 'deps' | 'semantic' | 'intelligence';

export interface TaskStatus {
    task_id: string;
    analyzer_type: AnalyzerType;
    state: TaskState;
    stage?: string;
    progress?: number;
    started_at?: string;
    completed_at?: string;
    error?: string;
    details?: Record<string, any>;
}

export type TaskState = 'not_run' | 'pending' | 'running' | 'completed' | 'failed';

export interface TaskResult {
    task_id: string;
    success: boolean;
    result?: any;
    error?: string;
    duration_ms?: number;
}

export interface AnalysisTask {
    task_id: string;
    repo_id: string;
    analyzer_type: AnalyzerType;
    config: Record<string, any>;
    state: TaskState;
    stage?: string;
    progress?: number;
    started_at?: string;
    completed_at?: string;
    error?: string;
}

// Helper to get task state color
export const getTaskStateColor = (state: TaskState): string => {
    switch (state) {
        case 'completed':
            return 'green';
        case 'running':
            return 'blue';
        case 'pending':
            return 'yellow';
        case 'failed':
            return 'red';
        case 'not_run':
            return 'gray';
        default:
            return 'gray';
    }
};

// Helper to get task state icon
export const getTaskStateIcon = (state: TaskState): string => {
    switch (state) {
        case 'completed':
            return 'CheckCircle';
        case 'running':
            return 'Loader';
        case 'pending':
            return 'Clock';
        case 'failed':
            return 'XCircle';
        case 'not_run':
            return 'Circle';
        default:
            return 'Circle';
    }
};
